### ISSUE TITLE
Subobject dependencies should be dynamic

### ISSUE THREAD

**Issue by philippjfr** [holoviz/param#410]
The `param.depends` decorator supports expressing dependencies on parameters on subobjects. However these dependencies are evaluated only on initialization meaning that if the object is not yet defined then it will error and if it is redefined later it will not set up that dependency. The following example should work but currently errors:

```python
import param

class A(param.Parameterized):
    
    value = param.Parameter()
    
class B(param.Parameterized):
    
    a = param.ObjectSelector()
    
    @param.depends('a.value', watch=True)
    def test_depends(self):
        print(self.a.value)

b = B()
b.a = A()
b.a.value = 1
```

**Comment by jbednar** [comment]
I guess whether this is a bug or a requested feature depends on your outlook on life. :-)  Seems like you're coming down  on the side of "possible feature" given that it wasn't part of the original design, but I agree that it _should_  behave properly in that case, making it also kinda be a bug...

**Comment by philippjfr** [comment]
Agreed, I clicked on feature request but then almost decided to change the label.

**Comment by ceball** [comment]
On Thu, May 14, 2020 at 8:54 PM James A. Bednar <notifications@github.com>
wrote:

> I guess whether this is a bug or a requested feature depends on your
> outlook on life. :-)
>
If it's code you wrote, someone's suggesting a feature enhancement. If it's
code someone else wrote, it's a bug report! :)
